---
title: "Introduction to ChIP-Seq and directory setup"
author: "Mary Piper, Radhika Khetani"
date: "Thursday, March 3rd, 2016"
---

Contributors: Mary Piper, Radhika Khetani

Approximate time: 30 minutes

## Learning Objectives

* understanding the experimental setup and design for ChIP-Seq experiments

## Introduction to ChIP-Seq
Chromatin immunoprecipitation (ChIP) experiments isolate the chromatin from a cell and immunoprecipitate (IP) DNA fragments bound to a protein of interest. In ChIP-Seq, the DNA fragments are sequenced, enriched regions of DNA or peaks are determined, and over-represented sequence motifs and functional annotations can be identified. 

![chipseq_overview](../img/chipseq_overall.png)

During this session we will be performing a complete workflow for ChIP-Seq analysis, starting with experimental design and generation of the raw sequencing reads and ending with functional enrichment analyses and motif discovery.

![chipseq_workflow_general](../img/chipseq_workflow_general.png)

## Experimental design and library preparation

Several steps are involved in the library preparation of protein-bound DNA fragments for sequencing: 

![exp_workflow](../img/chipseq_experimental_workflow.png)

1. After the chromatin is isolated from the cell, proteins are cross-linked to the DNA
2. The DNA is sheared into fragments (sonication)
3. A protein-specific antibody is used to immunoprecipitate the protein-bound DNA fragments
4. The crosslink is reversed and DNA purified
5. DNA fragments are size selected and amplified using PCR


Within the DNA fragments enriched for the regions binding to a protein of interest, only a fraction correspond to actual signal. The proportion of DNA fragments containing the actual binding site of the protein depends on the **number of active binding sites, the number of starting genomes, and the efficiency of the IP**. 

In addition, when performing ChIP-Seq, some sequences may appear enriched due to the following:

- Open chromatin regions are fragmented more easily than closed regions
- Repetitive sequences might seem to be enriched (inaccurately assessed copy numbers in the assembled genome)
- Uneven distribution of sequence reads across the genome

Therefore, proper controls are essential. A ChIP-Seq peak should be compared with the same region of the genome in a matched control.

![peaks](../img/chipseq_exp_peaks.png)

The same starting material should be divided to be used for both the protein-specific IP and the control. The control sample can be generated by one of the following recommended techniques: 

- No IP (input DNA) 
- No antibody ("mock IP")
- Non-specific antibody (IgG "mock IP")

![controls](../img/chipseq_exp_controls.png)

## Experimental Design

We will ultimately be comparing the binding profiles of [Nanog](www.nature.com/stemcells/2009/0909/090910/full/stemcells.2009.118.html) and [Pou5f1](www.nature.com/cr/journal/v12/n5/full/7290134a.html) (Oct4) from the [HAIB TFBS ENCODE collection](http://hgdownload.cse.ucsc.edu/goldenpath/hg19/encodeDCC/wgEncodeHaibTfbs/) that have been ChIP-sequenced with Illumina in H1 human embryonic stem cell line (h1-ESC) cells. **Nanog and Pou5f1 are both TFs that are involved in stem cell pluripotency.**

The experimental design used two replicates, with each replicate divided into three samples:

- Control input DNA
- Nanog IP
- Pou5f1 IP

<img src="../img/chipseq_exp_design.png" width=500>

To keep things manageable and allow algorithms to finish within a few minutes we will be using reads from 32.8 Mb of chromosome 12 (chr12:1,000,000-33,800,000) for all samples. 

The workflow for the ChIP-Seq analysis will start with quality control and alignment, proceed to peak calling and comparing peaks between samples, and finish with motif discovery and functional enrichment analyses. Similar to RNA-Seq, each step in the workflow will require the data to be in a specific type of standardized format.

![workflow](../img/chip_workflow_combined.png)

In the interest of time, we will not be using all samples in every step:

![workflow](../img/chipseq_analysis_workflow_samples.png)}}

## Set-up

Before we get started with the ChIP-Seq analysis, we need to set up our directory structure.

Login to Orchestra and start an interactive session with four cores:

`$ bsub -Is -n 4 -q interactive bash`

Change directories to the `ngs_course` directory:

`$ cd ~/ngs_course`

Create a `chipseq` directory and change directories into it:

```
$ mkdir chipseq

$ cd chipseq
```

Now let's setup the directory structure. In one command create folders for `data`and `results` _and_ within the `data` folder create folders for `untrimmed_fastq`, `trimmed_fastq`, and `reference_data`:

`$ mkdir -p data/untrimmed_fastq data/trimmed_fastq data/reference_data`

Within the `results` folder create folders for untrimmed_fastqc, trimmed_fastqc, and bowtie2:

`$ mkdir -p results/untrimmed_fastqc results/trimmed_fastqc results/bowtie2`

Now that we have the directory structure created, let's copy over the data to perform our quality control and alignment, including our fastq files and reference data files:

```
$ cd data

$ cp /groups/hbctraining/ngs-data-analysis-longcourse/chipseq/raw_fastq/*fastq untrimmed_fastq/

$ cp /groups/hbctraining/ngs-data-analysis-longcoursechipseq/reference_data/chr12* reference_data/
```

We will be using tools within the bcbio pipeline, so please make sure bcbio is in your PATH:

```
$ PATH=/opt/bcbio/centos/bin:$PATH
```
or within your `.bashrc` file:

```
export PATH=/opt/bcbio/centos/bin:$PATH
```
***
*This lesson has been developed by members of the teaching team at the [Harvard Chan Bioinformatics Core (HBC)](http://bioinformatics.sph.harvard.edu/). These are open access materials distributed under the terms of the [Creative Commons Attribution license](https://creativecommons.org/licenses/by/4.0/) (CC BY 4.0), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited.*

